crates/scsi/src/scsi/emulation/block_device.rs

```diff
@@ -0,0 +1,548 @@
+use std::{
+    convert::{TryFrom, TryInto},
+    fs::File,
+    io::{self, Read, Write},
+    os::unix::prelude::*,
+};
+
+use log::{debug, error, warn};
+use CmdError::DataIn;
+
+use super::{
+    command::{
+        parse_opcode, Cdb, Command, CommandType, ModePageSelection, ModeSensePageControl,
+        ParseError, ParseOpcodeResult, ReportLunsSelectReport, ReportSupportedOpCodesMode,
+        SenseFormat, VpdPage, OPCODES,
+    },
+    mode_page::ModePage,
+    response_data::{respond_report_luns, respond_standard_inquiry_data},
+    CmdError, CmdOutput, EmulatedTarget, LogicalUnit, Request,
+};
+use crate::scsi::{emulation::response_data::SilentlyTruncate, sense, TaskAttr};
+
+pub enum MediumRotationRate {
+    Unreported,
+    NonRotating,
+}
+
+pub struct BlockDevice {
+    file: File,
+    block_size: u32,
+    write_protected: bool,
+    rotation_rate: MediumRotationRate,
+}
+
+impl BlockDevice {
+    pub const fn new(file: File) -> Self {
```

Comment:
To clarify: Is `File` the correct abstraction level here? Or should we use some `Read`-like trait as backend here? That could allow for some more options: testing with super large (fake) backend, network backed implementations (not sure how great the idea is?).

Not exactly sure what level of synchronization is required for write support or a potential multithreading addition...